datastore/storagebackend/postgresql/gettsattrgroups.go

```diff
@@ -0,0 +1,326 @@
+	colValPtrs := make([]interface{}, len(cols)) // pointers to column values
+	for i := range colVals0 {
+		colValPtrs[i] = &colVals0[i]
+	}
+	// scan row into column value pointers
+	if err := rows.Scan(colValPtrs...); err != nil {
+		return nil, fmt.Errorf("rows.Scan() failed: %v", err)
+	}
+	// combine column names and -values into a map
+	colVals := map[string]interface{}{}
+	for i, col := range cols {
+		colVals[col] = colVals0[i]
+	}
+	// convert to a TSMetadata object
+	tsMdata, err := getTSMdata(colVals)
+	if err != nil {
+		return nil, fmt.Errorf("getTSMdata failed(): %v", err)
+	}
+
+	return tsMdata, nil
+}
+
+// getTSMdataEqual returns upon success (true, nil) iff tsmd1 and tsmd2 are equal wrt. goNames,
+// otherwise (false, nil). If an error occurs, (..., error) is returned.
+func tsMdataEqual(tsmd1, tsmd2 *datastore.TSMetadata, goNames []string) (bool, error) {
+	tp1 := reflect.ValueOf(tsmd1)
+	tp2 := reflect.ValueOf(tsmd2)
+
+	for _, goName := range goNames {
+		field1 := tp1.Elem().FieldByName(goName)
+		if !field1.IsValid() {
+			return false, fmt.Errorf("tp1.Elem().FieldByName(%s) returned invalid value", goName)
+		}
+		field2 := tp2.Elem().FieldByName(goName)
+		if !field2.IsValid() {
+			return false, fmt.Errorf("tp2.Elem().FieldByName(%s) returned invalid value", goName)
+		}
+
+		if !field1.Equal(field2) {
+			return false, nil // at least one difference found
+		}
+	}
+
+	return true, nil // no differences found
+}
+
+// getCombo gets a version of tsMdata1 where only the given attributes (in goNames) have been set
+// (all other attributes having the default values).
+// Returns (TSMetadata, nil) upon success, otherwise (..., error).
+func getCombo(tsMdata1 *datastore.TSMetadata, goNames []string) (*datastore.TSMetadata, error) {
+	tsMdata2 := datastore.TSMetadata{}
+	tp1 := reflect.ValueOf(tsMdata1)
+	tp2 := reflect.ValueOf(&tsMdata2)
+
+	for _, goName := range goNames {
+		field1 := tp1.Elem().FieldByName(goName)
+		if !field1.IsValid() {
+			return nil, fmt.Errorf("invalid field (1) (goName: %s)", goName)
+		}
+		field2 := tp2.Elem().FieldByName(goName)
+		if !field2.IsValid() {
+			return nil, fmt.Errorf("invalid field (2) (goName: %s)", goName)
+		}
+		if !field2.CanSet() {
+			return nil, fmt.Errorf("unassignable field (2) (goName: %s)", goName)
+		}
+
+		switch field1.Kind() {
+		case reflect.String:
+			field2.SetString(field1.String())
+		default:
+			return nil, fmt.Errorf("unsupported type: %v (goName: %s)", field1.Kind(), goName)
+		}
+	}
+
+	return &tsMdata2, nil
+}
+
+// getTSAttrGroupsIncInstances populates groups from cols such that each group contains all
+// instances that match a unique combination of database values corresponding to cols.
+// All attributes, including those in cols, are set to the actual values found in the database.
+// Returns nil upon success, otherwise error.
+func getTSAttrGroupsIncInstances(
+	db *sql.DB, cols []string, groups *[]*datastore.TSMdataGroup) error {
```

Comment:
Is there a specific reason why this is passed as an input parameters, and not just a return parameter? Same for the next function.

It might just be my lack of Go knowledge.

Reply:
In this case I guess it is slightly more readable to return the array rather than manipulate it via a pointer. Then the code itself makes it explicit that the array is supposed to be empty at the time of call. If the array was very big you would save some copying by operating via a pointer, but that is hardly the case here. So once again: code clarity over premature (and most likely insignificant) optimization! Change coming up ...